keras_nlp/models/roberta.py

```diff
@@ -0,0 +1,302 @@
+    encoder = keras_nlp.models.Roberta(
+        vocabulary_size=50265,
+        num_layers=12,
+        num_heads=12,
+        hidden_dim=768,
+        intermediate_dim=3072,
+        max_sequence_length=12
+    )
+
+    # Call encoder on the inputs.
+    input_data = {
+        "input_ids": tf.random.uniform(
+            shape=(1, 12), dtype=tf.int64, maxval=30522),
+        "input_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+    }
+    output = encoder(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.1,
+        max_sequence_length=512,
+        bos_token_index=0,
+        name=None,
+        trainable=True,
+    ):
+
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="input_ids"
+        )
+        input_mask = keras.Input(
+            shape=(None,), dtype="int32", name="input_mask"
+        )
+
+        # Embed tokens and positions.
+        token_and_position_embedding_layer = TokenAndPositionEmbedding(
+            vocabulary_size=vocabulary_size,
+            sequence_length=max_sequence_length,
+            embedding_dim=hidden_dim,
+            embeddings_initializer=_roberta_kernel_initializer(),
+            name="token_and_position_embeddings",
+        )
+        embedding = token_and_position_embedding_layer(token_id_input)
+
+        # Sum, normailze and apply dropout to embeddings.
+        x = keras.layers.LayerNormalization(
+            name="embeddings_layer_norm",
+            axis=-1,
+            epsilon=1e-5,  # Original paper uses this epsilon value
+            dtype=tf.float32,
+        )(embedding)
+        x = keras.layers.Dropout(
+            dropout,
+            name="embeddings_dropout",
+        )(x)
+
+        # Apply successive transformer encoder blocks.
+        for i in range(num_layers):
+            x = TransformerEncoder(
+                num_heads=num_heads,
+                intermediate_dim=intermediate_dim,
+                activation=lambda x: keras.activations.gelu(
+                    x, approximate=True
+                ),
+                dropout=dropout,
+                kernel_initializer=_roberta_kernel_initializer(),
+                name=f"""transformer_layer_{i}""",
+            )(x, padding_mask=input_mask)
+
+        # Construct the two RoBerta outputs.
+        output = x
+
+        # Instantiate using Functional API Model constructor
+        super().__init__(
+            inputs={
+                "input_ids": token_id_input,
+                # "segment_ids": segment_id_input,
+                "input_mask": input_mask,
+            },
+            outputs=output,
+            name=name,
+            trainable=trainable,
+        )
+        # All references to `self` below this line
+        self.token_embedding = (
```

Comment:
Does roberta do the same flip the embedding table trick for the MLM head that BERT does? Mainly asking for my own understanding.

Reply:
Yep it ties the embedding weights here: https://github.com/facebookresearch/fairseq/blob/acd9a53607d1e5c64604e88fc9601d0ee56fd6f1/fairseq/models/roberta/model.py#L555